qiskit/circuit/controlflow/builder.py

```diff
@@ -0,0 +1,386 @@
+            elif isinstance(bit, Clbit):
+                self.clbits.add(bit)
+            else:
+                raise TypeError(f"Can only add qubits or classical bits, but received '{bit}'.")
+
+    def build(
+        self, all_qubits: FrozenSet[Qubit], all_clbits: FrozenSet[Clbit]
+    ) -> "qiskit.circuit.QuantumCircuit":
+        """Build this scoped block into a complete :obj:`~QuantumCircuit` instance.
+
+        This will build a circuit which contains all of the necessary qubits and clbits and no
+        others.
+
+        The ``qubits`` and ``clbits`` arguments should be sets that contains all the resources in
+        the outer scope; these will be passed down to inner placeholder instructions, so they can
+        apply themselves across the whole scope should they need to.  The resulting
+        :obj:`.QuantumCircuit` will be defined over a (nonstrict) subset of these resources.  This
+        is used to let ``break`` and ``continue`` span all resources, even if they are nested within
+        several :obj:`.IfElsePlaceholder` objects, without requiring :obj:`.IfElsePlaceholder`
+        objects *without* any ``break`` or ``continue`` statements to be full-width.
+
+        Args:
+            all_qubits: all the qubits in the containing scope of this block.  The block may expand
+                to use some or all of these qubits, but will never gain qubits that are not in this
+                set.
+            all_clbits: all the clbits in the containing scope of this block.  The block may expand
+                to use some or all of these clbits, but will never gain clbits that are not in this
+                set.
+
+        Returns:
+            A circuit containing concrete versions of all the instructions that were in the scope,
+            and using the minimal set of resources necessary to support them, within the enclosing
+            scope.
+        """
+        from qiskit.circuit import QuantumCircuit
+
+        # There's actually no real problem with building a scope more than once.  This flag is more
+        # so _other_ operations, which aren't safe can be forbidden, such as mutating instructions
+        # that may have been built into other objects.
+        self._built = True
+
+        potential_qubits = all_qubits - self.qubits
+        potential_clbits = all_clbits - self.clbits
+
+        # We start off by only giving the QuantumCircuit the qubits we _know_ it will need, and add
+        # more later as needed.
+        out = QuantumCircuit(list(self.qubits), list(self.clbits))
+
+        for operation, qubits, clbits in self.instructions:
+            if isinstance(operation, InstructionPlaceholder):
+                operation, qubits, clbits = operation.concrete_instruction(all_qubits, all_clbits)
+                # We want to avoid iterating over the tuples unnecessarily if there's no chance
+                # we'll need to add bits to the circuit.  QuantumCircuit._append will already
```

Comment:
I suppose it might be impossible for `QuantumCircuit._append` to _not_ iterate over them, but in general, implementation remarks like this could get out of sync.

Reply:
Good point - removed in b24c37c50.